ecs-cli/modules/cli/compose/entity/service/service.go

```diff
@@ -364,12 +364,22 @@ func (s *Service) EntityType() types.Type {
 func (s *Service) buildCreateServiceInput(serviceName, taskDefName string) (*ecs.CreateServiceInput, error) {
 	launchType := s.Context().CommandConfig.LaunchType
 	cluster := s.Context().CommandConfig.Cluster
+	ecsParams := s.ecsContext.ECSParams
 
-	networkConfig, err := composeutils.ConvertToECSNetworkConfiguration(s.ecsContext.ECSParams)
+	networkConfig, err := composeutils.ConvertToECSNetworkConfiguration(ecsParams)
+	if err != nil {
+		return nil, err
+	}
+	placementConstraints, err := composeutils.ConvertToECSPlacementConstraints(ecsParams)
+	if err != nil {
+		return nil, err
+	}
+	placementStrategy, err := composeutils.ConvertToECSPlacementStrategy(ecsParams)
 	if err != nil {
 		return nil, err
 	}
 
+	// NOTE: this validation is not useful if called after GetOrCreateTaskDefinition()
```

Comment:
I'm curious about this- my reading of `ValidateFargateParams()` doesn't indicate it matters if its called before or after `GetOrCreateTaskDefinition()`- what am I missing?

Reply:
So there is nothing wrong with the `ValidateFargateParams()` method itself, but in the places it's being called from the top-level compose commands (in this case, `Create()`, we are calling `entity.GetOrCreateTaskDefinitions(s)` first (see the comment in `entity_helper.go` from the same [commit](https://github.com/aws/amazon-ecs-cli/pull/586/commits/275107a61e60bf6a05ae002bebdacaba1c4de77e)). In other words, due to the order-dependent nature of actual calls to ECS and the fact that we mock these in our unit tests, this Validation is not useful for validating the actual request, in this case, to CreateService, because bad parameters on the TaskDefinition itself will cause the server to return a validation error. E.g.:

```
// with the following inputs:

~/ecs/ecs-cli/testing/projects/test_fields$ cat ecs_params/ecs-params-task-placement.yml 
version: 1
task_definition:
run_params:
  task_placement:
    strategy:
      - field: memory
        type: binpack
      - field: attribute:ecs.availability-zone
        type: spread
    constraints:
      - expression: attribute:color =~ blue
        type: memberOf


~/ecs/ecs-cli/testing/projects/test_fields$ cat docker-compose.yml
version: '2'
services:
  wordpress:
    image: wordpress
    ports:
      - "80:80"
  mysql:
    image: mysql
    environment:
      MYSQL_ROOT_PASSWORD: foobar
```
ECS returns a ClientException on `RegisterTaskDefinition`:
```
~/ecs/ecs-cli/testing/projects/test_fields$ ecs-cli compose --ecs-params ecs_params/ecs-params-task-placement.yml up --launch-type FARGATE
ERRO[0000] Error registering task definition             error="ClientException: Fargate only supports network mode ‘awsvpc’.\n\tstatus code: 400, request id: c6f9aede-a562-11e8-bbee-ebd059c53f92" family=test_fields
ERRO[0000] Create task definition failed                 error="ClientException: Fargate only supports network mode ‘awsvpc’.\n\tstatus code: 400, request id: c6f9aede-a562-11e8-bbee-ebd059c53f92"
FATA[0000] ClientException: Fargate only supports network mode ‘awsvpc’.
	status code: 400, request id: c6f9aede-a562-11e8-bbee-ebd059c53f92 
```
I discovered this case when trying to test Fargate validation (since placement is not supported with Fargate launch type).